cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsCloudWatchService.java

```diff
@@ -74,35 +75,51 @@ public void addCloudWatchAlarmsForSystemFailures(List<CloudResource> instances,
         });
     }
 
-    public void deleteCloudWatchAlarmsForSystemFailures(CloudStack stack, String regionName, AwsCredentialView credentialView) {
+    public void deleteAllCloudWatchAlarmsForSystemFailures(CloudStack stack, String regionName, AwsCredentialView credentialView) {
         List<String> instanceIds = stack.getGroups().stream()
                 .flatMap(group -> group.getInstances().stream())
                 .map(CloudInstance::getInstanceId)
                 .collect(Collectors.toList());
+        LOGGER.info("Found ids for running instances: [{}]. Cloudwatch alarms for these instances will be deleted.", instanceIds);
+        List<String> deletedInstanceIds = stack.getGroups().stream()
+            .flatMap(group -> group.getDeletedInstances().stream())
+            .map(CloudInstance::getInstanceId)
+            .collect(Collectors.toList());
+        LOGGER.info("Found ids for deleted instances: [{}]. Deletion will be attempted on the cloudwatch alarms for these instances.", deletedInstanceIds);
+        instanceIds.addAll(deletedInstanceIds);
         deleteCloudWatchAlarmsForSystemFailures(stack, regionName, credentialView, instanceIds);
     }
 
     public void deleteCloudWatchAlarmsForSystemFailures(CloudStack stack, String regionName, AwsCredentialView credentialView, List<String> instanceIds) {
-        List<String> instanceIdsFromStack = stack.getGroups().stream()
+        if (instanceIds == null || instanceIds.isEmpty()) {
+            LOGGER.warn("No instance ids provided for cloudwatch alarm deletion. No alarms will be deleted.");
+        } else {
+            LOGGER.info("Deleting alarms for instance ids [{}]", instanceIds);
+            List<String> instanceIdsFromStack = stack.getGroups().stream()
```

Comment:
**suggestion:** From what I can tell, the body of this `else` is entirely to create the log warning statement, and it uses the `stack` only.

I think you can consider extracting this into a method like `private void logMissingInstances(CloudStack stack)`.

Reply:
It's mostly logging, but it also calls the next method in the call stack that's needed to actually do the delete:
```
deleteCloudWatchAlarmsForSystemFailures(regionName, credentialView, instanceIds);
```
I would prefer not extracting it out, since there's already a 3-4 overloaded method deep call stack, and I don't want to add another method to it unless it's really necessary.